cmd/tendermint/commands/run_node.go

```diff
@@ -48,31 +39,44 @@ func AddNodeFlags(cmd *cobra.Command) {
 	cmd.Flags().Bool("consensus.create_empty_blocks", config.Consensus.CreateEmptyBlocks, "Set this to false to only produce blocks when there are txs or when the AppHash changes")
 }
 
-// Users wishing to:
-//	* Use an external signer for their validators
-//	* Supply an in-proc abci app
-// should import github.com/tendermint/tendermint/node and implement
-// their own run_node to call node.NewNode (instead of node.NewNodeDefault)
-// with their custom priv validator and/or custom proxy.ClientCreator
-func runNode(cmd *cobra.Command, args []string) error {
-
-	genDocFile := config.GenesisFile()
-	genDoc, err := types.GenesisDocFromFile(genDocFile)
-	if err != nil {
-		return err
-	}
-	config.ChainID = genDoc.ChainID
-
-	// Create & start node
-	n := node.NewNodeDefault(config, logger.With("module", "node"))
-	if _, err := n.Start(); err != nil {
-		return fmt.Errorf("Failed to start node: %v", err)
-	} else {
-		logger.Info("Started node", "nodeInfo", n.Switch().NodeInfo())
-	}
+// FuncSignerAndApp takes a config and returns a PrivValidator and ClientCreator.
+// It allows other projects to make Tendermint binaries with custom signers and applications.
+type FuncSignerAndApp func(*cfg.Config) (types.PrivValidator, proxy.ClientCreator)
+
+// DefaultSignerAndApp is a default FuncSignerAndApp that returns a PrivValidatorFS
+// and a DefaultClientCreator using the relevant fields from the config.
+func DefaultSignerAndApp(config *cfg.Config) (types.PrivValidator, proxy.ClientCreator) {
+	privValidator := types.LoadOrGenPrivValidatorFS(config.PrivValidatorFile())
+	clientCreator := proxy.DefaultClientCreator(config.ProxyApp, config.ABCI, config.DBDir())
+	return privValidator, clientCreator
+}
 
-	// Trap signal, run forever.
-	n.RunForever()
+// NewRunNodeCmd returns the command that allows the CLI to start a
+// node. It can be used with a custom PrivValidator and in-process ABCI application.
+func NewRunNodeCmd(nodeFunc nm.NodeProvider) *cobra.Command {
```

Comment:
nodeFunc sounds like it's a (func) type.

If it's a function, how about nodeFn as a convention, or perhaps nodeCb, or in this case, provider?

Reply:
will change to nodeProvider

Reply:
xxxFn sounds good as a convention though. @zramsay want to add to coding repo? function vars should be followed by `Fn`

Reply:
I really like nodeFn